code/modules/keybindings/bindings_client.dm

```diff
@@ -0,0 +1,112 @@
+	//not entirely sure how byond commands interact in relation to lag
+	//don't want to kick people if a lag spike results in a huge flood of commands being sent
+	if(cache >= MAX_KEYPRESS_AUTOKICK)
+		if(!keysend_tripped)
+			keysend_tripped = TRUE
+			next_keysend_trip_reset = world.time + (2 SECONDS)
+		else
+			log_admin("Client [ckey] was just autokicked for flooding keysends; likely abuse but potentially lagspike.")
+			message_admins("Client [ckey] was just autokicked for flooding keysends; likely abuse but potentially lagspike.")
+			qdel(src)
+			return
+
+	///Check if the key is short enough to even be a real key
+	if(LAZYLEN(_key) > MAX_KEYPRESS_COMMANDLENGTH)
+		to_chat(src, "<span class='userdanger'>Invalid KeyDown detected! You have been disconnected from the server automatically.</span>")
+		log_admin("Client [ckey] just attempted to send an invalid keypress. Keymessage was over [MAX_KEYPRESS_COMMANDLENGTH] characters, autokicking due to likely abuse.")
+		message_admins("Client [ckey] just attempted to send an invalid keypress. Keymessage was over [MAX_KEYPRESS_COMMANDLENGTH] characters, autokicking due to likely abuse.")
+		qdel(src)
+		return
+	//offset by 1 because the buffer address is 0 indexed because the math was simpler
+	keys_held[current_key_address + 1] = _key
+	//the time a key was pressed isn't actually used anywhere (as of 2019-9-10) but this allows easier access usage/checking
+
+	keys_held[_key] = world.time
+
+	current_key_address = ((current_key_address + 1) % HELD_KEY_BUFFER_LENGTH)
+
+	var/movement = SSinput.movement_keys[_key]
+	if(!(next_move_dir_sub & movement) && !keys_held["Ctrl"])
+		next_move_dir_add |= movement
+
+	// Client-level keybindings are ones anyone should be able to do at any time
+	// Things like taking screenshots, hitting tab, and adminhelps.
+
+	switch(_key)
+		if("F1")
+			if(keys_held["Ctrl"] && keys_held["Shift"]) // Is this command ever used?
+				winset(src, null, "command=.options")
+			else
+				adminhelp()
+			return
+		if("F2", "O") // Screenshot. Hold shift to choose a name and location to save in
+			ooc()
+			return
+		if("F3", "T")
+			if(keys_held["Shift"])
+				mob.whisper_wrapper()
+			else
+				mob.say_wrapper()
+			return
+		if("F4", "M")
+			mob.me_wrapper()
+			return
+		if("L")
+			looc()
+			return
+		if("F11") // Toggles Fullscreen or Fits Viewport
+			if(keys_held["Ctrl"])
+				fit_viewport()
+			else
+				toggle_fullscreen()
+			return
+		if("F12") // Toggles minimal HUD
+			mob.button_pressed_F12()
+			return
+
+	if(holder)
+		holder.key_down(_key, src)
+	if(mob.focus)
+		mob.focus.key_down(_key, src)
+
+/client/verb/keyUp(_key as text)
+	set instant = TRUE
+	set hidden = TRUE
+	//Can't just do a remove because it would alter the length of the rolling buffer, instead search for the key then null it out if it exists
+	for(var/i in 1 to HELD_KEY_BUFFER_LENGTH)
+		if(keys_held[i] == _key)
+			keys_held[i] = null
+			break
```

Comment:
This needs to handle multiple instances of the same key in the ring buffer, as the KeyDown code does not. I get that that's an "error," but I'm almost certain you can produce it by manipulating client connections.